src/lib/optimizer/strategy/data_induced_predicate_rule.cpp

```diff
@@ -0,0 +1,130 @@
+using namespace expression_functional;  // NOLINT(build/namespaces)
+
+std::string DataInducedPredicateRule::name() const {
+  static const auto name = std::string{"DataInducedParameterRule"};
+  return name;
+}
+
+void DataInducedPredicateRule::_apply_to_plan_without_subqueries(
+    const std::shared_ptr<AbstractLQPNode>& lqp_root) const {
+  Assert(lqp_root->type == LQPNodeType::Root, "Rule needs root to hold onto");
+
+  // Adding dips inside visit_lqp might lead to endless recursions. Thus, we use visit_lqp to identify the
+  // reductions that we want to add to the plan, write them into data_induced_predicates and actually add them after
+  // visit_lqp.
+  std::vector<std::tuple<std::shared_ptr<JoinNode>, LQPInputSide, std::shared_ptr<PredicateNode>>>
+      data_induced_predicates;
+
+  const auto opposite_side = [](const auto side) {
+    return side == LQPInputSide::Left ? LQPInputSide::Right : LQPInputSide::Left;
+  };
+
+  const auto estimator = cost_estimator->cardinality_estimator->new_instance();
+  estimator->guarantee_bottom_up_construction();
+
+  visit_lqp(lqp_root, [&estimator, &opposite_side, &data_induced_predicates](const auto& node) {
+    if (node->type != LQPNodeType::Join) {
+      return LQPVisitation::VisitInputs;
+    }
+    const auto join_node = std::static_pointer_cast<JoinNode>(node);
+
+    for (const auto& join_predicate : join_node->join_predicates()) {
+      const auto predicate_expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(join_predicate);
+      DebugAssert(predicate_expression, "Expected BinaryPredicateExpression");
+      // Currently only equal predicates are supported.
+      if (predicate_expression->predicate_condition != PredicateCondition::Equals) {
+        continue;
+      }
+
+      // Since data induced predicates might be beneficial for both sides of the join, we create this helper lambda,
+      // which can deal with both sides.
+      const auto reduce_if_beneficial = [&](const auto selection_side) {
+        auto reduced_node = join_node->input(selection_side);
+        auto reducer_node = join_node->input(opposite_side(selection_side));
+
+        auto reducer_side_expression = predicate_expression->left_operand();
+        auto reduced_side_expression = predicate_expression->right_operand();
+
+        auto original_cardinality = estimator->estimate_cardinality(reduced_node);
+
+        /**
+         * We have to check whether the expression is actually evaluable on the reducer_node. If this is not the case
+         * we have to swap the sides. This might be the case, because it is not guaranteed, that the predicates are
+         * ordered according to the input sides.
+         */
+
+        if (!expression_evaluable_on_lqp(reducer_side_expression, *reducer_node)) {
+          std::swap(reduced_side_expression, reducer_side_expression);
+        }
+        DebugAssert(!expression_evaluable_on_lqp(reducer_side_expression, *join_node->input(selection_side)),
```

Comment:
Same question as above: Why is this a DebugAssert?

Reply:
See answer above